# A simple Python demo

## Readling `table.tsv` with Pandas

This notebook is a beginner-friendly tour of the most common pandas operations, using a real abundance table as example data:

- **rows** = taxa, identified by their full taxonomic lineage (e.g. `k__Bacteria,p__Firmicutes,...,g__Lactobacillus`)
- **columns** = samples (sequencing runs named `ERR...`)
- **values** = raw abundance counts

## 1. Reading the file

We load the TSV file into a pandas `DataFrame` (a table with labeled rows and columns), using the first column as row index, and preview the first rows.

In [1]:
import pandas as pd

df = pd.read_csv("table.tsv", sep="\t", index_col=0)
df.index.name = "Taxon"
df.head()

,ERR17070056,ERR17070057,ERR17070058,ERR17070059,ERR17070060,ERR17070061,ERR17070062,ERR17070063,ERR17070064,ERR17070065,...,ERR17070075,ERR17070076,ERR17070077,ERR17070078,ERR17070079,ERR17070080,ERR17070081,ERR17070082,ERR17070083,ERR17070084
Taxon,,,,,,,,,,,,,,,,,,,,,
u__unclassified,1503,1683,1107,1440,1702,1634,1403,1637,1042,1719,...,1934,2136,1890,1779,2162,243,1296,195,334,1250
"k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Lactobacillaceae,g__Lactobacillus",170195,252818,172036,163889,254344,241408,205856,226955,133054,268625,...,15829,58570,76659,191182,86619,295,217300,1703,29,133
"k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Lactobacillaceae,g__Pediococcus",1318,1964,1444,1249,1905,1855,1754,1803,1512,2256,...,982,6373,7164,10190,997,4,1930,37,1,2
"k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Leuconostocaceae,g__Leuconostoc",1500,43,22,34,36,36,1060,248,55,32,...,82662,1641,83534,27,6570,8,617,275,2,0
"k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Leuconostocaceae,g__Weissella",46,52,32,41,42,59,582,63,52,57,...,2037,184,33983,5654,59088,1,583,20,0,0


**🧑‍💻 Syntax:**
- `import pandas as pd` makes the library available under the short name `pd`, the universal convention.
- `pd.read_csv(...)` reads tabular text files; `sep="\t"` says columns are separated by tabs, and `index_col=0` uses the first column (the taxon names) as row labels instead of data.
- `df.head()` returns the first 5 rows; a value on the last line of a cell is displayed automatically in notebooks (no `print` needed).

**🔍 The output:** the DataFrame preview shows taxa as rows and samples as columns. With 29 samples the display is too wide, so pandas hides the middle columns (`...`).

## 2. Anatomy of a DataFrame

Before manipulating data, let's inspect its dimensions and labels: how many taxa, how many samples, and what they are called.

In [4]:
print("Shape (rows, columns):", df.shape)
print("\nFirst 3 samples:", list(df.columns[:3]))
print("\nFirst taxon:", list(df.index[:1]))
print("\nSecond taxon:", list(df.index[1:2]))

print("\nData type of the values:", df.dtypes.unique())

Shape (rows, columns): (102, 29)

First 3 samples: ['ERR17070056', 'ERR17070057', 'ERR17070058']

First taxon: ['u__unclassified']

Second taxon: ['k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Lactobacillaceae,g__Lactobacillus']

Data type of the values: [dtype('int64')]


**🧑‍💻 Syntax:**
- `df.shape`, `df.columns`, `df.index` and `df.dtypes` are *attributes* (no parentheses): they describe the object instead of performing an action.
- `df.columns[:3]` slices the first 3 labels; `list(...)` converts them to a plain Python list for nicer printing.
- `\n` inside a string starts a new line when printed.

**🔍 The output:** the table has **102 taxa x 29 samples**, and all values are integers (`int64`). Sample names follow the `ERR...` run accession format; taxon names are full lineage strings.

## 3. Selecting rows, columns and single values

We extract one sample (a column), one taxon (a row), and a single count, using both *label-based* (`loc`) and *position-based* (`iloc`) selection.

In [5]:
# One column (sample) -> a Series
sample = df["ERR17070056"]
print("Counts for ERR17070056, first 5 taxa:")
print(sample.head())

# One value by position: row 2, column 1
print("\ndf.iloc[2, 1] =", df.iloc[2, 1])
print("  (taxon:", df.index[2][:60] + "...)")
print("  (sample:", df.columns[1], ")")

# One value by label
print("\nBy label:", df.loc["u__unclassified", "ERR17070056"])

Counts for ERR17070056, first 5 taxa:
Taxon
u__unclassified                                                                                   1503
k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Lactobacillaceae,g__Lactobacillus    170195
k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Lactobacillaceae,g__Pediococcus        1318
k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Leuconostocaceae,g__Leuconostoc        1500
k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Leuconostocaceae,g__Weissella            46
Name: ERR17070056, dtype: int64

df.iloc[2, 1] = 1964
  (taxon: k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__L...)
  (sample: ERR17070057 )

By label: 1503


**🧑‍💻 Syntax:**
- `df["name"]` selects a single column and returns a `Series` (a labeled vector) - the workhorse object of pandas.
- `df.iloc[row, column]` selects by **integer position** (0-based), `df.loc[row_label, column_label]` by **label**. Both accept a single value or slices.
- Long labels can be sliced like strings: `df.index[2][:60]` takes the first 60 characters.

**🔍 The output:** `df.iloc[2, 1]` returns 1964: the value at the *3rd row, 2nd column* (Pediococcus in ERR17070057). `df.loc["u__unclassified", "ERR17070056"]` returns 1503 by name instead. Positions depend on row order, labels don't: prefer `loc` when you know names, `iloc` when you know positions.

## 4. Filtering with boolean masks

A *boolean mask* is a Series of `True`/`False` values, one per row. It is the standard way to filter data: we keep only the taxa that satisfy a condition.

In [6]:
# Which taxa have more than 1000 counts in sample ERR17070056?
mask = sample > 1000
print("Taxa with >1000 counts:", mask.sum(), "out of", len(mask))
print(sample[mask])

# Combine conditions: abundant in BOTH of two samples (& = and, | = or)
both = (df["ERR17070056"] > 1000) & (df["ERR17070057"] > 1000)
print("\nAbundant (>1000) in both samples:", both.sum())
print(df.index[both].str[:70])

Taxa with >1000 counts: 12 out of 102
Taxon
u__unclassified                                                                                                     1503
k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Lactobacillaceae,g__Lactobacillus                      170195
k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Lactobacillaceae,g__Pediococcus                          1318
k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Leuconostocaceae,g__Leuconostoc                          1500
k__Bacteria,p__Firmicutes,c__Bacilli,o__Staphylococcales,f__Staphylococcaceae,g__Staphylococcus                    20277
k__Bacteria,p__Firmicutes,c__Bacilli,o__Exiguobacterales,f__Exiguobacteraceae,g__Exiguobacterium                    5319
k__Bacteria,p__Firmicutes,c__Bacilli,o__Bacillales,f__Bacillaceae,g__Bacillus                                       1500
k__Bacteria,p__Proteobacteria,c__Gammaproteobacteria,o__Enterobacterales,f__Enterobacteriaceae,g__Enterobacte

**🧑‍💻 Syntax:**
- Comparisons like `sample > 1000` are applied *element-wise* and return a boolean Series.
- `sample[mask]` keeps only rows where the mask is `True`.
- Conditions are combined with `&` (and), `|` (or), `~` (not) - **not** Python's `and`/`or`, which don't work element-wise. Each condition must be wrapped in parentheses.
- `mask.sum()` counts the `True` values (True = 1); `len(mask)` gives the total number of rows.

**🔍 The output:** only 12 of the 102 taxa exceed 1000 counts in that sample, and just 4 are above 1000 in both samples - the community is dominated by a few abundant taxa (mostly *Lactobacillus*), while the rest are rare.

## 5. Sorting

We rank the taxa of one sample from most to least abundant.

In [7]:
sample.sort_values(ascending=False).head()

Taxon
k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Lactobacillaceae,g__Lactobacillus                  170195
k__Bacteria,p__Firmicutes,c__Bacilli,o__Staphylococcales,f__Staphylococcaceae,g__Staphylococcus                20277
k__Bacteria,p__Proteobacteria,c__Gammaproteobacteria,o__Pseudomonadales,f__Moraxellaceae,g__Acinetobacter      11749
k__Bacteria,p__Firmicutes,c__Bacilli,o__Exiguobacterales,f__Exiguobacteraceae,g__Exiguobacterium                5319
k__Bacteria,p__Proteobacteria,c__Gammaproteobacteria,o__Pseudomonadales,f__Pseudomonadaceae,g__Pseudomonas      4890
Name: ERR17070056, dtype: int64

**🧑‍💻 Syntax:**
- `sort_values()` returns a **new, sorted** Series and leaves the original untouched (pandas methods are not in-place by default).
- `ascending=False` sorts from largest to smallest; chaining `.head()` afterwards is a common idiom to get a "top N".

**🔍 The output:** the top 5 taxa of this sample: *Lactobacillus* dominates with ~170k counts, more than 8x the second-ranked *Staphylococcus* (~20k).

## 6. Simple math: totals and relative abundances

We sum counts per sample (library sizes) and per taxon (overall abundance), then convert raw counts to percentages so samples of different depth become comparable.

In [8]:
sample_totals = df.sum(axis=0)   # one total per sample (column sums)
taxon_totals = df.sum(axis=1)    # one total per taxon (row sums)

print("5 smallest libraries:")
print(sample_totals.sort_values().head())

# Relative abundance (%) of each taxon in one sample
rel = sample / sample_totals["ERR17070056"] * 100
print("\nTop 5 taxa by relative abundance in ERR17070056:")
rel.sort_values(ascending=False).head().round(2)

5 smallest libraries:
ERR17070084      2744
ERR17070082     17179
ERR17070080     22714
ERR17070083     47870
ERR17070071    138317
dtype: int64

Top 5 taxa by relative abundance in ERR17070056:


Taxon
k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Lactobacillaceae,g__Lactobacillus                  74.04
k__Bacteria,p__Firmicutes,c__Bacilli,o__Staphylococcales,f__Staphylococcaceae,g__Staphylococcus                8.82
k__Bacteria,p__Proteobacteria,c__Gammaproteobacteria,o__Pseudomonadales,f__Moraxellaceae,g__Acinetobacter      5.11
k__Bacteria,p__Firmicutes,c__Bacilli,o__Exiguobacterales,f__Exiguobacteraceae,g__Exiguobacterium               2.31
k__Bacteria,p__Proteobacteria,c__Gammaproteobacteria,o__Pseudomonadales,f__Pseudomonadaceae,g__Pseudomonas     2.13
Name: ERR17070056, dtype: float64

**🧑‍💻 Syntax:**
- `sum(axis=0)` collapses **rows** (result: one value per column); `sum(axis=1)` collapses **columns** (one value per row). A trick: `axis=0` is the direction along which row labels disappear.
- Arithmetic between a Series and a scalar (`sample / total * 100`) is applied element-wise.
- `.round(2)` rounds the display to 2 decimals.

**🔍 The output:** library sizes are very uneven - the smallest sample has ~2.7k total counts, the largest ~358k (a >100x difference). Relative abundance fixes this: in sample ERR17070056, *Lactobacillus* is ~74% of the community regardless of how deep it was sequenced.

## 7. Working with text: extract the genus from the lineage

Row labels are long lineage strings. Using pandas' string methods, we extract just the genus name (the part after `g__`) into a new, readable label.

In [9]:
genus = df.index.to_series().str.extract(r"g__([^,]+)")[0].fillna("unclassified").rename("genus")

print("Original label:\n ", df.index[1][:80], "...\n")
print("Extracted genus:\n ", genus.iloc[1])
print("\nFirst 8 genera:", list(genus.iloc[:8]))

Original label:
  k__Bacteria,p__Firmicutes,c__Bacilli,o__Lactobacillales,f__Lactobacillaceae,g__L ...

Extracted genus:
  Lactobacillus

First 8 genera: ['unclassified', 'Lactobacillus', 'Pediococcus', 'Leuconostoc', 'Weissella', 'Oenococcus', 'Streptococcus', 'Lactococcus']


**🧑‍💻 Syntax:**
- `df.index.to_series()` converts the index to a Series so we can use the `.str` accessor, which unlocks string operations on all labels at once (the result keeps the taxon labels as its index, so it stays aligned with `df`).
- `extract(r"g__([^,]+)")` uses a *regular expression*: `g__` matches the literal prefix, `([^,]+)` captures all following characters up to the next comma. The `r"..."` prefix means "raw string" (backslashes are not escaped).
- `extract` returns a DataFrame with one column per capture group, so `[0]` takes the first (only) group. `fillna("unclassified")` replaces missing values (rows without `g__`, like `u__unclassified`).

**🔍 The output:** each lineage is reduced to a short genus name, e.g. `Lactobacillus`. This makes labels readable and enables grouping by genus in the next step.

## 8. Grouping: aggregate rows by genus

Several rows can share the same genus. `groupby` implements the *split-apply-combine* pattern: split rows into groups, apply a summary function (here `sum`) to each group, combine the results.

In [10]:
by_genus = df.groupby(genus).sum()
print("Rows before groupby:", len(df), "-> after:", len(by_genus))

top10 = by_genus.sum(axis=1).sort_values(ascending=False).head(10)
print("\nTop 10 genera by total abundance (all samples):")
print(top10)

Rows before groupby: 102 -> after: 100

Top 10 genera by total abundance (all samples):
genus
Lactobacillus      3306159
Lactococcus         836381
Staphylococcus      469170
Weissella           378415
Leuconostoc         322343
Exiguobacterium     165224
Enterobacter        146900
Enterococcus        131185
Streptococcus       116391
Pseudomonas         107293
dtype: int64


**🧑‍💻 Syntax:**
- `df.groupby(genus)` groups the rows of `df` by the labels in the `genus` Series (which is row-aligned with `df`); `.sum()` then adds up the counts of each group, sample by sample.
- The result is a new DataFrame whose index is the unique genera.

**🔍 The output:** 102 taxa collapse into 100 genera (a couple of genera, e.g. *uncultured*, appear under multiple families and are merged). *Lactobacillus* is by far the most abundant genus overall (~3.3M counts), followed by *Lactococcus* and *Staphylococcus*.

## 9. Reshaping: from wide to long ("tidy") format

The table is *wide* (one column per sample). Many analysis tools prefer a *long* table with one row per (taxon, sample, count) combination. `melt` performs this conversion.

In [11]:
long = df.reset_index().melt(id_vars="Taxon", var_name="Sample", value_name="Count")
print("Wide shape:", df.shape, "-> Long shape:", long.shape)
long.head(3)

# Filtering works on long tables too: nonzero records only
print("\nRows with Count > 0:", (long["Count"] > 0).sum(), "of", len(long))

Wide shape: (102, 29) -> Long shape: (2958, 3)

Rows with Count > 0: 2009 of 2958


**🧑‍💻 Syntax:**
- `reset_index()` moves the index (taxon names) into a regular column called `Taxon`.
- `melt(id_vars=...)` keeps `Taxon` fixed and stacks all sample columns into two new columns: `Sample` (former column name) and `Count` (value).
- `long["Count"]` selects a column of the long DataFrame like any other.

**🔍 The output:** 102 x 29 becomes 2958 x 3 (= 102 taxa x 29 samples rows). 2009 of the 2958 records (~68%) are nonzero: the table is sparser than the wide view suggests - about one third of the (taxon, sample) combinations are zero.

## 10. Exporting results

Any DataFrame or Series can be written back to disk. We save the top-10 genera computed in step 8 to a CSV file.

In [12]:
top10.rename("total_counts").to_csv("top_genera.csv", index_label="genus")

# Read it back to verify
pd.read_csv("top_genera.csv")

,genus,total_counts
0,Lactobacillus,3306159
1,Lactococcus,836381
2,Staphylococcus,469170
3,Weissella,378415
4,Leuconostoc,322343
5,Exiguobacterium,165224
6,Enterobacter,146900
7,Enterococcus,131185
8,Streptococcus,116391
9,Pseudomonas,107293


**🧑‍💻 Syntax:**
- `rename("total_counts")` gives the Series a name, which becomes the column header in the file.
- `to_csv("file.csv", index_label="genus")` writes the file next to the notebook and names the index column `genus`.
- Reading it back with `pd.read_csv` is a quick sanity check.

**🔍 The output:** the file `top_genera.csv` now contains the ranking table and can be opened in Excel or shared. Delete it freely - it is regenerated each time this cell runs.

## Recap and next steps

In this tutorial we covered the core pandas workflow:

| Step | Key functions |
|---|---|
| Read | `pd.read_csv(sep, index_col)` |
| Inspect | `.shape`, `.head()`, `.columns`, `.index`, `.dtypes` |
| Select | `df["col"]`, `.loc`, `.iloc` |
| Filter | boolean masks, `&`, `|`, `~` |
| Sort | `.sort_values(ascending=False)` |
| Math | `.sum(axis=...)`, element-wise arithmetic, `.round()` |
| Text | `.str.extract()`, `.str[:n]` |
| Group | `.groupby(...).sum()` |
| Reshape | `.reset_index()`, `.melt()` |
| Write | `.to_csv()` |

Ideas to practice further: compute per-sample diversity (number of nonzero taxa), correlations between samples (`df.corr()`), or explore `describe()` and `value_counts()`. See `explore_table.ipynb` for the plotting counterpart of this tutorial.